File: Structural_Patterns/Flyweight/cpp/FlyweightFactory.cpp
/*!
 *  \file       FlyweightFactory.cpp
 *  \brief      
 *  
 */


#include "FlyweightFactory.hpp"

#include <iostream>


namespace dive_into_design_patterns
{
    FlyweightFactory::FlyweightFactory(std::initializer_list<SharedState> shared_states)
    {
        for (const SharedState &ss : shared_states)
        {
            m_flyweights.insert(std::make_pair(getKey(ss), Flyweight(&ss)));
        }
    }

    std::string FlyweightFactory::getKey(const SharedState &ss) const
    {
        return ss.getBrand() + "_" + ss.getModel() + "_" + ss.getColor();
    }

    Flyweight FlyweightFactory::getFlyweight(const SharedState &shared_state)
    {
        std::string key = getKey(shared_state);
        if (m_flyweights.find(key) == m_flyweights.end())
        {
            std::cout << "FlyweightFactory: cannot find a flyweight, creating a new one." << std::endl;
            m_flyweights.insert(std::make_pair(key, Flyweight(&shared_state)));
        }
        else
        {
            std::cout << "FlyweightFactory: Reusing existing flyweight." << std::endl;
        }
        
        return m_flyweights.at(key);
    }

    void FlyweightFactory::listFlyweights() const
    {
        size_t count = m_flyweights.size();
        std::cout << "FlyweightFactory: I have " << count << " flyweights:" << std::endl;
        for (const auto& [key, flyweight] : m_flyweights)
        {
            std::cout << key << std::endl;
        }
    }
}

File: Structural_Patterns/Flyweight/python/FlyweightFactory.py
# File name     : FlyweightFactory.py
# Description   : 
#
#

from typing import Dict

from Flyweight import Flyweight

class FlyweightFactory:
    '''
    The Flyweight Factory creates and manages the Flyweight objects.
    It ensures that flyweights are shared correctly. When the client requests a flyweight,
    the factory either returns an existing instance or creates a new one, if it doesn't exist yet.
    '''
    _flyweights: Dict[str, Flyweight] = {}

    def __init__(self, initial_flyweights: Dict):
        for state in initial_flyweights:
            self._flyweights[self.get_key(state)] = Flyweight(state)

    def get_key(self, state: Dict):
        return "_".join(sorted(state))
    
    def get_flyweight(self, shared_state: Dict):
        key = self.get_key(shared_state)

        if not self._flyweights.get(key):
            print("FlyweightFactory: Cannot find a flyweight, creating new one.")
            self._flyweights[key] = Flyweight(shared_state)
        else:
            print("FlyweightFactory: Reusing existing flyweight.")

        return self._flyweights[key]
    
    def list_flyweights(self):
        count = len(self._flyweights)
        print(f"FlyweightFactory: I have {count} flyweights:")
        print("\n".join(map(str, self._flyweights.keys())), end="")